plugins/tokens/plugin.go

```diff
@@ -51,6 +51,76 @@ func createQueryHandler(mapper Mapper, queryPrefix string) app.AbciQueryHandler
 	return createAbciQueryHandler(mapper, queryPrefix)
 }
 
+const (
+	MaxUnlockItems = 10
+)
+
+func EndBlocker(ctx sdk.Context, timelockKeeper timelock.Keeper, swapKeeper swap.Keeper) {
+	if !sdk.IsUpgrade(sdk.SecondSunsetFork) {
+		return
+	}
+	logger := bnclog.With("module", "tokens")
+	logger.Info("unlock the time locks", "blockHeight", ctx.BlockHeight())
+
+	iterator := timelockKeeper.GetTimeLockRecordIterator(ctx)
+	defer iterator.Close()
+	i := 0
+	failedCount := 0
```

Comment:
In cosmos-sdk, we also have a limit for failures. Do we also need such a strategy here?

Reply:
`TimeUnlock` is a very simple function. It shouldn't fail.
It failed and I think we need to do manual analysis

![Uploading image.png…]()